### DISCUSSION TITLE
Gamepad and touchscreen input???

### DISCUSSION THREAD

**Discussion by squiblez** [bjornbytes/lovr#599] [Q&A] [answered]
Are these possible without adding it to the source myself? I decided to use lovr to experiment in 3D programming before i moved it to my own engine in mono but it took off and I'd actually prefer to keep it in lovr. Gamepad input is almost required for this, touchscreen would be nice.

Also thanks to everyone involved, lovr is the first and only engine I've ever worked with other than my own and enjoyed.

**Comment by bjornbytes** [comment, marked as answer]
Hi!  Neither are built in right now.  There have been 2 previous efforts for joystick input:

- https://github.com/bjornbytes/lovr/pull/222
- https://github.com/Rabios/lovr-joystick

OpenXR does support gamepads, but only a single one, and only with some headsets.

**Reply by squiblez** [reply]
I couldn't get either one of them to work on Linux, but [https://github.com/Rabios/lovr-joystick](url) did give me an idea. I've started writing a soft driver in C# by reading bytes directly from Unix device files(since I'm working on Linux) and once its completed I'll adapt it into Lua for use as a lovr module. Anyone curious about this can check it out in my repos.

**Comment by rempas** [comment]
Couldn't a library that handles inputs like `SDL_Input` can be used?

**Reply by bjornbytes** [reply]
Yes, it could.  LÖVR doesn't use SDL right now, and it's kind of a huge dependency to add and would require removing GLFW and the existing platform code.  So it might be more work, compared to exposing gamepad/touchscreen input using the structure that already exists.

**Reply by rempas** [reply]
> Yes, it could. LÖVR doesn't use SDL right now, and it's kind of a huge dependency to add and would require removing GLFW and the existing platform code. So it might be more work, compared to exposing gamepad/touchscreen input using the structure that already exists.

Couldn't we add it manually and just use `SDL_Input` manually and let LOVR do its job the way it wants with GLFW?

**Reply by bjornbytes** [reply]
Oh sure, you could do that

**Reply by rempas** [reply]
> Oh sure, you could do that

Sounds great! Do you know of any tutorials about how to work with external libraries and *mix* them to LOVR?

**Reply by bjornbytes** [reply]
Here's some documentation about how native libraries can integrate with lovr, along with some example plugins that could be used as a reference

https://lovr.org/docs/Plugins

**Reply by rempas** [reply]
> Here's some documentation about how native libraries can integrate with lovr, along with some example plugins that could be used as a reference
> 
> https://lovr.org/docs/Plugins

Thank you so much! I'll take a look at it! I hope GLFW is replaced with SDL before version 1.0 where the API will become stable. Godot also recently started using `SDL_Input`.

**Comment by iamgreaser** [comment]
Got inspired to look into how GLFW handles touch... which is to say, it doesn't do anything outside of treating it like a mouse pointer, so multitouch is not supported. Yeah, OK, I'm going to see how hard it is to actually support SDL3. I at least have a blank window, haven't got the OpenXR side of things working yet.

**Reply by bjornbytes** [reply]
There are a few existing examples of using SDL

https://github.com/bjornbytes/lovr/compare/dev...snowkittykira:lovr-fork:sdl3
https://github.com/bjornbytes/lovr/pull/938

I don't think SDL needs to interact with OpenXR much

**Reply by iamgreaser** [reply]
Thanks for the links. Seems pretty doable.

I'd argue in favour of splitting off the non-GLFW code for Win32 and Linux just so we don't have everywhere doing `#ifdef LOVR_USE_GLFW` but other than that the first one does seem like a better approach out of the two.
